File: backend/agent.js
import dotenv from "dotenv";
dotenv.config();

const API_KEY  = process.env.GROQ_API_KEY || process.env.GOOGLE_API_KEY;
const GROQ_URL = "https://api.groq.com/openai/v1/chat/completions";
const MODEL    = "meta-llama/llama-4-scout-17b-16e-instruct";

const SYSTEM = `You are HCA's product assistant for Hari Chand Anand & Co. — an industrial sewing machinery company in India.
You help salespeople and customers find the right machine and price instantly.

Rules:
1. Only use data returned by your tools. Never guess specs or prices.
2. For prices: ALWAYS call get_price. If the first attempt returns an error, retry with variations:
   - Strip brand prefix (e.g. "DUKEJIA DY-1201" → try "DY-1201")
   - Try without spaces (e.g. "DY 1201" → try "DY-1201")
   - Try the key/code from search results if available
   Make at least 2 attempts before saying price is unavailable.
3. When a user asks about a machine AND its price, call search_machines and get_price in the same turn.
4. For deal pricing: say "For the best deal, I'll connect you with our sales team."
5. Be direct and brief — salespeople are on live calls.
6. If a machine is not in the catalog after searching, say so clearly.`;

const TOOLS = [
  {
    type: "function",
    function: {
      name: "search_machines",
      description: "Search HCA machine catalog by keyword, brand, or category. Use for finding machines, listing options, recommendations.",
      parameters: {
        type: "object",
        properties: {
          q:        { type: "string", description: "Keyword e.g. embroidery, lockstitch, button hole, overlock" },
          brand:    { type: "string", description: "DUKE or DUKEJIA" },
          category: { type: "string", description: "Category e.g. embroidery, lockstitch" }
        }
      }
    }
  },
  {
    type: "function",
    function: {
      name: "get_machine_details",
      description: "Get complete specs for one specific machine. Use when user asks about a specific model or wants full details.",
      parameters: {
        type: "object",
        properties: {
          id: { type: "string", description: "Machine ID from search results e.g. duke-dk-1201-embroidery-machine" }
        },
        required: ["id"]
      }
    }
  },
  {
    type: "function",
    function: {
      name: "get_price",
      description: "Get quoted price for a machine model. Use whenever user asks about price, cost, or rate.",
      parameters: {
        type: "object",
        properties: {
          machine_model: { type: "string", description: "Machine model name e.g. DY-1201, DUKE R5, DUKE R9" }
        },
        required: ["machine_model"]
      }
    }
  }
];

const BASE = process.env.API_BASE || "http://localhost:3001";

async function runTool(name, args) {
  try {
    if (name === "search_machines") {
      const p = new URLSearchParams();
      if (args.q)        p.set("q", args.q);
      if (args.brand)    p.set("brand", args.brand);
      if (args.category) p.set("category", args.category);
      const res  = await fetch(`${BASE}/api/machines?${p}`);
      const data = await res.json();
      return data.slice(0, 8).map(m => ({
        id:          m.id,
        brand:       m.brand,
        model:       m.model,
        category:    m.category,
        operation:   m.operation,
        description: m.description,
        catalogUrl:  m.media?.catalogUrl
      }));
    }

    if (name === "get_machine_details") {
      const res = await fetch(`${BASE}/api/machines/${args.id}`);
      if (!res.ok) return { error: "Machine not found in catalog" };
      const m = await res.json();
      return {
        id: m.id, brand: m.brand, model: m.model,
        category: m.category, operation: m.operation,
        description: m.description, spec: m.spec,
        features: m.features, application: m.application,
        catalogUrl: m.media?.catalogUrl
      };
    }

    if (name === "get_price") {
      const res = await fetch(`${BASE}/api/price/${encodeURIComponent(args.machine_model)}`);
      if (!res.ok) return { error: "Price not available for this model" };
      return await res.json();
    }

    return { error: `Unknown tool: ${name}` };

  } catch (e) {
    return { error: e.message };
  }
}

export async function chat(userMessage, history = []) {
  if (!API_KEY) throw new Error("No API key found. Set GROQ_API_KEY in backend/.env");

  // Build message list: system + history (OpenAI format) + new user message
  const messages = [
    { role: "system", content: SYSTEM },
    ...history,
    { role: "user", content: userMessage }
  ];

  let finalText = "";
  const cards = [];

  for (let round = 0; round < 5; round++) {
    const res = await fetch(GROQ_URL, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${API_KEY}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({ model: MODEL, messages, tools: TOOLS, tool_choice: "auto", parallel_tool_calls: false })
    });

    if (!res.ok) {
      const errText = await res.text();
      throw new Error(`Groq API error ${res.status}: ${errText}`);
    }

    const data    = await res.json();
    const message = data.choices[0].message;
    messages.push(message);

    const toolCalls = message.tool_calls || [];

    if (toolCalls.length === 0) {
      finalText = message.content || "";
      break;
    }

    // Execute all tool calls
    for (const tc of toolCalls) {
      const name   = tc.function.name;
      const args   = JSON.parse(tc.function.arguments || "{}");
      const result = await runTool(name, args);

      // Collect structured card data for the UI
      if (name === "search_machines" && Array.isArray(result) && result.length > 0) {
        cards.push({ type: "machine_list", machines: result });
      } else if (name === "get_machine_details" && result && !result.error) {
        cards.push({ type: "machine_detail", machine: result });
      } else if (name === "get_price" && result && !result.error) {
        cards.push({ type: "price", model: result.model, quote_price_inr: result.quote_price_inr, currency: result.currency });
      }

      messages.push({
        role:         "tool",
        tool_call_id: tc.id,
        content:      JSON.stringify(result)
      });
    }
  }

  // Keep last 20 messages (excluding system) for next turn
  const updatedHistory = messages.slice(1).slice(-20);
  return { answer: finalText, history: updatedHistory, cards };
}

File: python-agent/agent.py
import json
import os
from langchain_groq import ChatGroq
from langchain_core.messages import AIMessage, ToolMessage
from langgraph.prebuilt import create_react_agent
from langgraph.checkpoint.postgres import PostgresSaver
from psycopg_pool import ConnectionPool

from tools import search_machines, get_machine_details, get_price, get_projects, get_installation_stats, search_catalog_docs

SYSTEM_PROMPT = """You are the HCA Sales Assistant for Hari Chand Anand & Co. — an industrial sewing machinery company in India.
Contact: tech@grouphca.com

## RULE #1 — MANDATORY, NO EXCEPTIONS
You have ZERO knowledge of what machines, brands, models, or specs HCA carries.
Every answer about catalog, brands, models, pricing, or installations MUST come from a tool call.
If you have not called a tool yet this turn, you cannot answer — call the tool first.

## Step 1 — Before writing any response, classify the query

Ask yourself: "Is this about machines, brands, models, categories, prices, installations, factory lines, or technical specs?"
- If YES → find the right tool below and call it NOW. Do not respond yet.
- If NO (e.g. greeting, general question) → respond directly.

## Step 2 — Tool decision table

| Query type | Tool to call |
|---|---|
| What brands / machines do you carry? | search_machines(q="") |
| Machines by a specific brand | search_machines(brand="<brand>") |
| Machines by type or category | search_machines(category="<type>") |
| Specific model or keyword search | search_machines(q="<keyword>") |
| Full specs of one machine | get_machine_details(id) using id from search results |
| Price / cost / rate of any model | get_price(machine_model="...") — retry with stripped model name if 404 |
| Factory lines / complete production setups | get_projects() |
| Installations / city presence / geographic reach | get_installation_stats() |
| Technical specs from catalog PDF (stitch length, needle, motor, thread) | search_catalog_docs(query="...", model="...") |

### Price retry logic
- Attempt 1: full name as given
- Attempt 2: strip brand prefix
- Attempt 3: normalise separators (spaces → hyphens)
- Attempt 4: use `key` field from search_machines results

## Step 3 — Respond after tool results are in hand

- Write ONE brief sentence intro (e.g. "Here are our embroidery machines:")
- Data renders as cards — do NOT list machine names, specs, or prices in text
- Never paste catalog URLs or video links in text — they are on the cards
- For bulk/volume pricing: "Contact our sales team at tech@grouphca.com for bulk pricing."

## What you must never do
- State brand names, model counts, or spec values without a tool call that returned them
- Say "full range available" or "various options available" from memory
- Answer a catalog question without calling a tool

## Website navigation (no tool needed)
- Machines page: machines.html
- Projects page: projects.html
- Installation map: map.html
- 3D Exhibition: exhibition.html"""

TOOLS = [search_machines, get_machine_details, get_price, get_projects, get_installation_stats, search_catalog_docs]

# ── LLM ─────────────────────────────────────────────────────────────────
llm = ChatGroq(
    model="qwen/qwen3-32b",
    api_key=os.getenv("GROQ_API_KEY"),
    max_retries=2,
    temperature=0,
)

# ── Postgres checkpointer ────────────────────────────────────────────────
DB_URI = os.getenv("DATABASE_URL")

_pool = None
_graph = None


def get_graph():
    """Lazy-init: build graph + pool once, reuse across requests."""
    global _pool, _graph
    if _graph is not None:
        return _graph

    _pool = ConnectionPool(
        conninfo=DB_URI,
        max_size=20,
        kwargs={"autocommit": True},
    )
    checkpointer = PostgresSaver(_pool)
    checkpointer.setup()   # creates langgraph_checkpoint tables (idempotent)

    _graph = create_react_agent(
        model=llm,
        tools=TOOLS,
        prompt=SYSTEM_PROMPT,
        checkpointer=checkpointer,
    )
    return _graph


# ── Cards extraction ─────────────────────────────────────────────────────
def extract_cards(messages: list) -> list:
    """
    Parse tool messages from the final LangGraph state and build
    the same card structures the frontend expects.
    Only processes messages from the current turn (after the last HumanMessage).
    """
    from langchain_core.messages import HumanMessage

    # Slice to current turn only so cards don't accumulate across turns
    last_human_idx = 0
    for i, msg in enumerate(messages):
        if isinstance(msg, HumanMessage):
            last_human_idx = i
    messages = messages[last_human_idx:]

    cards = []

    # Map tool_call_id → tool name from AI messages
    tool_name_map: dict[str, str] = {}
    for msg in messages:
        if isinstance(msg, AIMessage) and msg.tool_calls:
            for tc in msg.tool_calls:
                tool_name_map[tc["id"]] = tc["name"]

    for msg in messages:
        if not isinstance(msg, ToolMessage):
            continue
        tool_name = tool_name_map.get(msg.tool_call_id, "")
        try:
            content = json.loads(msg.content) if isinstance(msg.content, str) else msg.content

            if tool_name == "search_machines" and isinstance(content, list) and content:
                machines = [m for m in content if "error" not in m and "message" not in m]
                if machines:
                    cards.append({"type": "machine_list", "machines": machines})

            elif tool_name == "get_machine_details" and isinstance(content, dict) and "error" not in content:
                cards.append({"type": "machine_detail", "machine": content})

            elif tool_name == "get_price" and isinstance(content, dict) and "error" not in content:
                cards.append({
                    "type":            "price",
                    "model":           content.get("model"),
                    "quote_price_inr": content.get("quote_price_inr"),
                    "currency":        content.get("currency", "INR"),
                })

            elif tool_name == "get_projects" and isinstance(content, list) and content:
                projects = [p for p in content if "error" not in p and "message" not in p]
                if projects:
                    cards.append({"type": "project_list", "projects": projects})
        except Exception:
            pass

    return cards


# ── Public interface ─────────────────────────────────────────────────────
def run_chat(message: str, thread_id: str) -> dict:
    """
    Invoke the graph for one user message.
    Returns { answer, cards, history_turns }.
    Thread state (full history) is loaded/saved automatically via PostgresSaver.
    """
    from langchain_core.messages import HumanMessage

    graph = get_graph()
    config = {"configurable": {"thread_id": thread_id}}

    result = graph.invoke(
        {"messages": [HumanMessage(content=message)]},
        config=config,
    )

    messages      = result["messages"]
    answer        = messages[-1].content if messages else ""
    cards         = extract_cards(messages)
    history_turns = sum(1 for m in messages if isinstance(m, AIMessage))

    return {
        "answer":        answer,
        "cards":         cards,
        "history_turns": history_turns,
    }
